kolibri/core/content/upgrade.py

```diff
@@ -0,0 +1,111 @@
+"""
+A file to contain specific logic to handle version upgrades in Kolibri.
+"""
+import logging
+import os
+
+from kolibri.core.content.models import ChannelMetadata
+from kolibri.core.content.models import ContentNode
+from kolibri.core.content.utils.annotation import annotate_content
+from kolibri.core.content.utils.channel_import import FutureSchemaError
+from kolibri.core.content.utils.channel_import import import_channel_from_local_db
+from kolibri.core.content.utils.channel_import import InvalidSchemaVersionError
+from kolibri.core.content.utils.channels import get_channel_ids_for_content_database_dir
+from kolibri.core.content.utils.paths import get_content_database_dir_path
+from kolibri.core.content.utils.paths import get_content_database_file_path
+from kolibri.core.upgrade import VersionUpgradeBase
+
+logger = logging.getLogger(__name__)
+
+
+class ImportExternalContentDBs(VersionUpgradeBase):
+    # Only bother doing this if we are moving from
+    # a version of Kolibri before we imported
+    # content databases.
+    OLD_VERSION = "<0.6.0"
+
+    def upgrade(self):
+        """
+        If we are potentially moving from a version of Kolibri that did not import its content data,
+        scan through the content database folder for all channel content databases,
+        and pull the data from each database if we have not already imported it.
+        """
+        channel_ids = get_channel_ids_for_content_database_dir(
+            get_content_database_dir_path()
+        )
+        for channel_id in channel_ids:
+            if not ChannelMetadata.objects.filter(id=channel_id).exists():
+                try:
+                    import_channel_from_local_db(channel_id)
+                    annotate_content(channel_id)
+                except (InvalidSchemaVersionError, FutureSchemaError):
+                    logger.warning(
+                        "Tried to import channel {channel_id}, but database file was incompatible".format(
+                            channel_id=channel_id
+                        )
+                    )
```

Comment:
This is not a blocking comment, but I'll add my usual inheritance-averse reaction:

I think this logic would be better represented as instances of a `VersionUpgrade` class rather than subclasses of a `VersionUpgradeBase` class.

The main reason is that with subclasses, there's an implicit assumption that subclass can "extend" the API or change behaviors as desired. Having class instances instead makes the API and contract fully defined and immutable, which as far as I can tell would better serve the purpose of the `run_upgrades` function.

For example:

```python
from kolibri.core.upgrade import VersionUpgrade

def importExternalContentDBs(self):
    """
    If we are potentially moving from a version of Kolibri that did not import its content data,
    scan through the content database folder for all channel content databases,
    and pull the data from each database if we have not already imported it.
    """
    channel_ids = get_channel_ids_for_content_database_dir(
        get_content_database_dir_path()
    )
    for channel_id in channel_ids:
        if not ChannelMetadata.objects.filter(id=channel_id).exists():
            try:
                import_channel_from_local_db(channel_id)
                annotate_content(channel_id)
            except (InvalidSchemaVersionError, FutureSchemaError):
                logger.warning(
                    "Tried to import channel {channel_id}, but database file was incompatible".format(
                        channel_id=channel_id
                    )
                )

def getUpgrades():
    return [
        VersionUpgrade(func=importExternalContentDBs, old_version="<0.6.0"),
    ]
```

Reply:
Fair comment - I just thought I'd get away with it because I restrained myself from using metaclasses :)

Reply:
Have now refactored to rely on instantiation, with a convenience decorator to just simply apply it to a function to make it into an upgrade operation.